### DISCUSSION TITLE
RFC: Make DatePicker/TimePicker/Calendar date-library-agnostic

### DISCUSSION THREAD

**Discussion by yesmeck** [ant-design/ant-design#36887] [RFCs]
Previous discussions https://github.com/ant-design/ant-design/issues/5677, https://github.com/ant-design/ant-design-mobile/issues/1217.

## Motivation

`moment` do have some downsides like it's [mutable value](https://github.com/moment/moment/issues/1754) and [un-modular](https://github.com/moment/moment/issues/2373). And  it's road to modular seems very slow. I also found that the moment team is writing another date library called [luxon](https://github.com/moment/luxon). It would be a big improvement if we can make our date related components date-library-agnostic.

## Detailed design

Create a `rc-date-util` package. It only contains one line:

```javascript
export default from 'rc-moment-adapter';
```

`rc-moment-adapter` wraps all APIs which date components used.

All date related components use `rc-date-util` instead of using `moment` directly. 

People can change date library through webpack's alias:

```javascript
// webpack.config.js

module.exports = {
  resolve: {
    alias: {
      'rc-date-util': 'rc-date-fns-adapter',
    }
  }
}
```

Browserify users can use [aliasify](https://www.npmjs.com/package/aliasify):

```json
{
  "aliasify": {
    "aliases": {
       "rc-date-util": "rc-date-fns-adapter",
    }
  }
}
```

After aliasing `rc-moment-adapter` to `rc-date-fns-adapter`,  all dates pass to callbacks should change to native date object as well.

```javascript
const onChange = (date)  => {
  console.log(date); // Native date object.
}

<DatePicker onChange={onChange} />
```

This change won't effects any users who don't want to change moment to other date library.

## Drawbacks

Rely on build tool config.

**Comment by andriijas** [comment]
WIll the adapters be small in size? another concern with moment related to unmodular is the huge size. What are the most common operations on time and date in antd that brings in the need for a library? can we detect those and see if its better to have adapters or just write our own antd date helpers? i guess we dont want to maintain own code managing dates but if the result is less code than adapters it should be considered.

**Comment by yutingzhao1991** [comment]
>```
> const onChange = (date)  => {
>   console.log(date); // Native date object.
> }
> ```

Should `date` keep to use moment or other date library?

We can add a common method in rc-[xxx]-adapter like `parseFromNativeDate` for this.

**Comment by yesmeck** [comment]
The adapter should be very small. I don't think we will repalce moment with our own date helper, because we don't want break any exists code, but we can have a `antd-vanilla-date-adapter` or something whatever name which don't reply on any date library.

**Comment by yesmeck** [comment]
> Should date keep to use moment or other date library?

`date` type changes only if `rc-date-util` is replaced with other adapter.

**Comment by andriijas** [comment]
Sounds great,  thanks for making moment opt-in! Appreciate it!

**Comment by hlehmann** [comment]
I really like this feature, any update on this ?

**Comment by otakustay** [comment]
Depending on build tools to implement type adapters is not a good solution, why don't we just make `dateUtil` as a prop of date pickers with a `defaultProps` value as official `rc-date-utils`

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/issues/10437 is a better proposal

**Comment by hk-ng** [comment]
#10437 is closed. Should we reconsider date-library-agnostic solution here?

**Comment by maxcan** [comment]
Strong proponent of luxon.  Its combines moment clean, easy to use API with the immutability of date-fns.  

there is at least one abstraction library available which may make this easier: https://github.com/dmtrKovalenko/date-io
